### PR TITLE
feat: Showing build and deployment duration 

### PR DESCRIPTION
**Pull request by ShashwatDevtron**
# Description

These changes will enable one section on build and deployment history pages which will show time taken to complete the build and deployment.

fix #https://github.com/devtron-labs/devtron/issues/2927

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

All tests have been performed manually by visiting build and deployment history pages under app details and observing the progress section while building and deploying app respectively .

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/app/details/cicdHistory/TriggerDetails.tsx [hunk 1]
```diff
@@ -87,17 +90,25 @@ export const TriggerDetails = React.memo(
     },
 )
 
-const Finished = React.memo(({ status, finishedOn, artifact }: FinishedType): JSX.Element => {
+const Finished = React.memo(({ status, startedOn, finishedOn, artifact }: FinishedType): JSX.Element => {
```
Comment (shivani170) [inline src/components/app/details/cicdHistory/TriggerDetails.tsx:93]: Could you please add Finished prop type? It is missing here

File: src/components/app/details/cicdHistory/TriggerDetails.tsx [hunk 2]
```diff
@@ -87,17 +90,25 @@ export const TriggerDetails = React.memo(
...
     return (
         <div className="flex column left">
             <div className={`${status} fs-14 fw-6 ${TERMINAL_STATUS_COLOR_CLASS_MAP[status.toLowerCase()] || 'cn-5'}`}>
                 {status && status.toLowerCase() === 'cancelled' ? 'ABORTED' : status}
             </div>
+
```
Comment (shivani170) [inline src/components/app/details/cicdHistory/TriggerDetails.tsx:99]: remove this line

File: src/components/app/details/cicdHistory/TriggerDetails.tsx [hunk 3]
```diff
@@ -87,17 +90,25 @@ export const TriggerDetails = React.memo(
...
             <div className="flex left">
                 {finishedOn && finishedOn !== ZERO_TIME_STRING && (
-                    <time className="cn-7 fs-12 mr-12">
-                        {moment(finishedOn, 'YYYY-MM-DDTHH:mm:ssZ').format(Moment12HourFormat)}
-                    </time>
+                    <>            
+                        <time className="dc__vertical-align-middle">
+                            {moment(finishedOn, 'YYYY-MM-DDTHH:mm:ssZ').format(Moment12HourFormat)}
+                        </time>
+                        <div className="dc__bullet mr-6 ml-6"></div>
```
Comment (shivani170) [inline src/components/app/details/cicdHistory/TriggerDetails.tsx:106]: Can use <div className="dc__bullet mr-6 ml-6"/>  if no children.

File: src/components/app/details/cicdHistory/TriggerDetails.tsx [hunk 4]
```diff
@@ -131,9 +143,10 @@ const WorkerStatus = React.memo(({ message, podStatus, stage }: WorkerStatusType
 })
 
 const ProgressingStatus = React.memo(
-    ({ status, message, podStatus, stage, type }: ProgressingStatusType): JSX.Element => {
+    ({ status, startedOn, message, podStatus, stage, type }: ProgressingStatusType): JSX.Element => {
```
Comment (shivani170) [inline src/components/app/details/cicdHistory/TriggerDetails.tsx:146]: Type is missing here

File: src/components/app/details/cicdHistory/TriggerDetails.tsx [hunk 5]
```diff
@@ -131,9 +143,10 @@ const WorkerStatus = React.memo(({ message, podStatus, stage }: WorkerStatusType
...
         const [aborting, setAborting] = useState(false)
         const [abortConfirmation, setAbortConfiguration] = useState(false)
+        const [durationStr, setDurationStr] = useState('');
```
Comment (shivani170) [inline src/components/app/details/cicdHistory/TriggerDetails.tsx:149]: Type is missing

File: src/components/app/details/cicdHistory/TriggerDetails.tsx [hunk 6]
```diff
@@ -145,6 +158,20 @@ const ProgressingStatus = React.memo(
         } else if (stage !== 'DEPLOY') {
             abort = () => cancelPrePostCdTrigger(pipelineId, triggerId)
         }
+
+        useEffect(() => {
+            setDurationStr(formatDurationDiff(startedOn, Date()))
```
Comment (shivani170) [inline src/components/app/details/cicdHistory/TriggerDetails.tsx:163]: Please for built in formatter to formate time

File: src/components/common/helpers/time.ts [hunk 7]
```diff
@@ -35,3 +35,8 @@ export function handleUTCTime(ts: string, isRelativeTime = false) {
     }
     return timestamp;
 }
+
+export const formatDurationDiff = (startedOn: string, finishedOn: string) => {
```
Comment (shivani170) [inline src/components/common/helpers/time.ts:39]: We can  use inbuilt formatters

Comment (ShashwatDevtron) [reply src/components/common/helpers/time.ts:39]: Yes the built in formatter was not formatting time in the way that we wanted it to be displayed, so I created my own formatter, Sohel  checked the code and told me to do so.

File: src/components/app/details/cicdHistory/TriggerDetails.tsx (hunk 6 above)
Comment (ShashwatDevtron) [reply src/components/app/details/cicdHistory/TriggerDetails.tsx:163]: ### Yes the built in formatter was not formatting time in the way that we wanted it to be displayed, so I created my own formatter, Sohel  checked the code and told me to do so.

Comment (ShashwatDevtron) [reply src/components/app/details/cicdHistory/TriggerDetails.tsx:163]: Yes the built in formatter was not formatting time in the way that we wanted it to be displayed, so I created my own formatter, Sohel  checked the code and told me to do so.

File: src/components/app/details/cicdHistory/TriggerDetails.tsx (hunk 3 above)
Comment (ShashwatDevtron) [reply src/components/app/details/cicdHistory/TriggerDetails.tsx]: Ok sure, Done that

File: src/components/app/details/cicdHistory/TriggerDetails.tsx (hunk 4 above)
Comment (ShashwatDevtron) [reply src/components/app/details/cicdHistory/TriggerDetails.tsx:146]: Yes I have given the type in ProgressingStatusType.

File: src/components/app/details/cicdHistory/TriggerDetails.tsx (hunk 1 above)
Comment (ShashwatDevtron) [reply src/components/app/details/cicdHistory/TriggerDetails.tsx:93]: I have given the type in FinishedType

File: src/components/app/details/cicdHistory/TriggerDetails.tsx (hunk 5 above)
Comment (ShashwatDevtron) [reply src/components/app/details/cicdHistory/TriggerDetails.tsx]: Sure, done that

File: src/components/app/details/cIDetails/ciDetails.scss [hunk 8]
```diff
@@ -67,6 +67,13 @@
                     .inprogress-status-color {
                       color: var(--inprogress-orange);
                     }
+                    .abort-status-color {
```
[the remark this answers is not in the document]
Comment (ShashwatDevtron) [reply src/components/app/details/cIDetails/ciDetails.scss]: Done the changes as mentioned

File: src/components/app/details/cicdHistory/TriggerDetails.tsx (hunk 2 above)
Comment (ShashwatDevtron) [reply src/components/app/details/cicdHistory/TriggerDetails.tsx]: Done.

File: src/components/common/helpers/time.ts [hunk 9]
```diff
@@ -35,3 +35,8 @@ export function handleUTCTime(ts: string, isRelativeTime = false) {
...
+   const diff : moment.Duration = moment.duration(moment(finishedOn).diff(moment(startedOn)))
```
Comment (shivani170) [inline src/components/common/helpers/time.ts:40]: Rename it more precisely

**Review by shivani170 — APPROVED** [review]

File: src/components/common/helpers/time.ts (hunk 9 above)
Comment (sohel-rp) [reply src/components/common/helpers/time.ts:40]: we've kept it in helper so name is given accordingly :)

**Review by sohel-rp — APPROVED** [review]

**Pull request merged**